File: LeetCodeProblems/All/cpp/572.subtree-of-another-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 
/*----------SOLUTION----------*/

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q){
        if(p==NULL and q==NULL){
            return true;
        }
        if(p==NULL or q==NULL){
            return false;
        }
        if(p->val==q->val){
            bool leftMatch=isSameTree(p->left, q->left);
            bool rightMatch=isSameTree(p->right, q->right);
            return leftMatch and rightMatch;
        }else{
            return false;
        }
    }
    bool isSubtree(TreeNode* s, TreeNode* t) {
        if(s==NULL and t==NULL){
            return true;
        }
        if(s==NULL or t==NULL){
            return false;
        }
        if(isSameTree(s,t)){
            return true;
        }
        bool leftMatch=isSubtree(s->left,t);
        bool rightMatch=isSubtree(s->right,t);
        return leftMatch or rightMatch;
    }
};

File: LeetCode/Tag/Depth-first-Search/js/572.subtree-of-another-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */
var isSubtree = function(root, subRoot) {
    if(!root){
      return false;
    }

    if(root.val === subRoot.val){
      if(checkIfSameTree(root,subRoot)){
        return true;
      }
    }
    return isSubtree(root.left,subRoot) || isSubtree(root.right,subRoot);
};

const checkIfSameTree = function(root,subRoot){
  if(!root && !subRoot)
    return true;
  if(!subRoot && root)
    return false;
  if(!root && subRoot)
    return false;
  if(root.val !== subRoot.val)
    return false;

  return checkIfSameTree(root.left, subRoot.left) && checkIfSameTree(root.right, subRoot.right);

}
